src/pudl/analysis/timeseries_cleaning.py

```diff
@@ -1560,6 +1569,200 @@ def filter_missing_values(
+
+    Each row in this dataframe will contain a reference ID and month, and a simulation
+    ID and month. The reference ID/month points to a month of reported data where
+    a high ratio of values were flagged for imputation, and the simulation ID/month
+    points to a month of reported data with no flagged values. The flagged values
+    in the reference month will be used to simulate flagged values in the simulation
+    month.
+    """
+
+    reference_id_col: Series[Any]
+    reference_month: Series[pa.dtypes.DateTime]
+    simulation_id_col: Series[Any]
+    simulation_month: Series[pa.dtypes.DateTime]
+
+
+@pa.check_types
+def _merge_imputed(
+    aligned_df: DataFrame[AlignedTimeseriesDataFrame],
+    matrix: DataFrame[TimeseriesMatrix],
+    flags: DataFrame[TimeseriesMatrix],
+) -> pd.DataFrame:
+    """Helper function to melt imputed timeseries matrix and merge back on input asset."""
+    # Melt imputed matrix/flag matrix to long format
+    imputed_df = melt_imputed_timeseries_matrix(matrix, flags)
+
+    # Merge back on core table
+    return aligned_df.merge(
+        imputed_df.rename(columns={"value_col": "imputed_value_col"}),
+        on=["id_col", "datetime"],
+        how="left",
+    )
+
+
+@pa.check_types
+def _add_simulated_flag_col(
+    imputed_df: DataFrame[AlignedTimeseriesDataFrame],
+    simulation_df: DataFrame[SimulationDataFrame],
+) -> DataFrame[AlignedTimeseriesDataFrame]:
+    """Return a modified ``imputed_df`` with a column indicating which rows should be flagged for simulation.
+
+    This will find all flagged values from a reference month and apply the flag
+    pattern to a simulation month. The flag pattern is determined by calculating the
+    hour of the month for each flagged (how many hours is this after the start of the
+    month), and flagging the corresponding hour in the simulation month. Reference
+    months are chosen by finding months with a relatively high rate of imputation,
+    while simulation months have no values which were flagged for imputation.
+
+    Args:
+        imputed_df: Production DataFrame with imputed values, which is used to find
+            sections with high rates of imputation.
+        simulation_df: DataFrame with reference and simulation months.
+
+    Returns:
+        DataFrame which contains all ID/datetime pairs that should be flagged for simulated imputation.
+    """
+    # Add a column to both dataframes, which contains the start date of the month
+    # In the ``datetime`` column.
+    simulation_df["period"] = simulation_df["reference_month"].dt.to_period("M")
+    imputed_df["period"] = imputed_df["datetime"].dt.to_period("M")
+
+    # Merge on month and ID to get a DataFrame with all hours in the reference months
+    simulation_df = imputed_df.merge(
+        simulation_df,
+        left_on=["id_col", "period"],
+        right_on=["reference_id_col", "period"],
+        how="inner",
```

Comment:
Please add a merge validation for legibility and error checking.

Reply:
Done.